File: android-native/KyouNoOgatore/app/src/main/java/jp/ogatore/kyouno/KyonoAsyncImage.kt
package jp.ogatore.kyouno

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import androidx.compose.foundation.Image
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.net.HttpURLConnection
import java.net.URL

// 見た目パリティ第2弾(TASK-C2-2026-07-26-visual-parity-round2.md §2 動画サムネイル): 新規ライブラリ
// 依存(Coil等)を追加する実績がプロジェクト内に無いため、「標準のAsyncImage相当」として
// produceState+HttpURLConnectionによる素朴な非同期読み込みをここに実装する。読み込み失敗時は
// Web版onerror="this.style.visibility='hidden'"と同じく何も表示しない(崩れたアイコンを見せない)。
//
// TASK: オフライン用ディスクキャッシュ(ThumbnailCache)を読み書き経路に挟む。弱電波下で検索/
// 再生リスト画面を開くたびに毎回ネット取得していた問題への対策(§前回コメント参照)。
// context.cacheDir/thumbnails/配下に保存し、context.filesDir(kyono-store.json=記録データ本体)とは
// 完全に分離する。書き込み済みならネットワーク状況に関わらず常にキャッシュを優先してよい
// (サムネイルURLはvideoId+size単位で不変のため鮮度管理/期限切れは不要)。
@Composable
fun KyonoAsyncImage(url: String, modifier: Modifier = Modifier, contentScale: ContentScale = ContentScale.Crop) {
    val context = LocalContext.current
    val cache = remember(context) { ThumbnailCache(File(context.cacheDir, "thumbnails")) }
    val bitmap by produceState<Bitmap?>(initialValue = null, key1 = url) {
        value = runCatching {
            withContext(Dispatchers.IO) {
                val key = ThumbnailCache.keyFor(url)
                val bytes = cache.read(key) ?: run {
                    val conn = URL(url).openConnection() as HttpURLConnection
                    conn.connectTimeout = 8000
                    conn.readTimeout = 8000
                    conn.doInput = true
                    conn.connect()
                    val fetched = conn.inputStream.use { it.readBytes() }
                    cache.write(key, fetched)
                    fetched
                }
                BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
            }
        }.getOrNull()
    }
    bitmap?.let {
        Image(
            bitmap = it.asImageBitmap(),
            contentDescription = null,
            contentScale = contentScale,
            modifier = modifier,
        )
    }
}

// index.html videos.js/app-search.js相当: YouTube動画IDからサムネイルURLを組み立てる。
// 一覧(検索・再生リスト)はmqdefault、個別強調(ツアーミニチュア等)はhqdefaultをWeb版で使い分けている。
fun youtubeThumbUrl(videoId: String, size: String = "mqdefault"): String =
    "https://i.ytimg.com/vi/$videoId/$size.jpg"

File: ios-native/KyouNoOgatore/KyouNoOgatore/KyonoAsyncImage.swift
//
//  KyonoAsyncImage.swift
//  KyouNoOgatore
//
//  見た目パリティ第2弾(TASK-C2-2026-07-26-visual-parity-round2.md §2 動画サムネイル): iOS標準の
//  AsyncImage(タスク文の推奨どおり・新規ライブラリ依存なし)を薄くラップする。読み込み失敗時は
//  Web版onerror="this.style.visibility='hidden'"と同じく何も表示しない(崩れたアイコンを見せない)。
//
//  TASK: オフライン用ディスクキャッシュ(ThumbnailCache)を読み書き経路に挟むため、SwiftUI標準の
//  AsyncImageから自前のURLSession+.task(id:)実装に切り替えた(AsyncImageはキャッシュ挙動を
//  差し込めないため)。Cachesディレクトリ配下のthumbnails/に保存し、Documentsディレクトリ
//  (kyono-store.json=記録データ本体)とは完全に分離する。書き込み済みならネットワーク状況に
//  関わらず常にキャッシュを優先してよい(サムネイルURLはvideoId+size単位で不変)。
//
//  TASK-C2-2026-07-29-testflight-feedback-d.md D1: 実機でサムネイルが1枚も出ない不具合。
//  NSLogでの実測(検索結果24行・全行 `.task` 発火ゼロ)で確定: bodyが`Group { if let uiImage {...} }`
//  のとき、uiImage==nilの間はGroupの中身が完全に空になり、その空Groupに付けた`.task`が
//  一度も発火しない(SwiftUIが中身の無いGroupをレンダーツリーにmountしない)。`Color.clear`を
//  常に実体のあるアンカーにして`.overlay`で条件付き内容を載せる形に変えたところ、同じ実測で
//  全行の`.task`発火・画像取得(200 OK)を確認した。

import SwiftUI
import UIKit

struct KyonoAsyncImage: View {
    let url: String
    var contentMode: ContentMode = .fill

    @State private var uiImage: UIImage?

    private static let cache: ThumbnailCache = {
        let base = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
        let dir = base.appendingPathComponent("thumbnails", isDirectory: true)
        return ThumbnailCache(cacheDir: dir)
    }()

    var body: some View {
        Color.clear
            .overlay {
                if let uiImage {
                    // F2(TASK-C2-2026-07-29-inspection-upgrade.md): D1(サムネイル全滅)は行数だけを
                    // 見るテストでは素通りしていた(画像が1枚も無くても行は描画されるため)。画像が
                    // 実際に読み込めて描画されたときにだけ付く識別子を用意し、UIテスト側で
                    // 「行が存在する」だけでなく「その中に画像が実在する」ことまで確認できるようにする。
                    Image(uiImage: uiImage).resizable().aspectRatio(contentMode: contentMode)
                        .accessibilityIdentifier("kyonoThumbnailLoaded")
                }
            }
            .task(id: url) {
                // produceState(key1=url)相当: urlが変わったら古い画像を即座に消してから読み込み直す
                // (画面再利用時に別動画の古いサムネイルが一瞬見えるのを防ぐ)。
                uiImage = nil
                uiImage = await Self.loadImage(urlString: url)
            }
    }

    private static func loadImage(urlString: String) async -> UIImage? {
        let key = ThumbnailCache.key(forURL: urlString)
        if let cached = cache.read(key: key), let image = UIImage(data: cached) {
            return image
        }
        guard let requestURL = URL(string: urlString) else { return nil }
        guard let (data, _) = try? await URLSession.shared.data(from: requestURL) else { return nil }
        guard let image = UIImage(data: data) else { return nil }
        cache.write(key: key, data: data)
        return image
    }
}

// index.html videos.js/app-search.js相当: YouTube動画IDからサムネイルURLを組み立てる。
func youtubeThumbUrl(_ videoId: String, size: String = "mqdefault") -> String {
    "https://i.ytimg.com/vi/\(videoId)/\(size).jpg"
}
